### DISCUSSION TITLE
Is there any timeline to Wix4.0?

### DISCUSSION THREAD

**Discussion by ChristianSauer** [wixtoolset/issues#6929] [Questions]
I am honestly tired of finding problems / limitations in wix and reading half-decade old reports which tell "will be fixed in wix4.0" Either release it or shut the project down so I can move on. Both is ok, but this zombie is annoying

**Comment by robmen** [comment]
You can use WiX v4.0 now.

**Reply by ChristianSauer** [reply]
Nope, you cannot really with confidence.

Look at the website: https://wixtoolset.org/releases/ 
Latest stable is [WiX v3.11.2](https://wixtoolset.org/releases/v3.11.2/stable)

In the news section there is mention of a preview from 2011: "WiX Toolset v4-preview.0 available"

if I google "wix 4.0" I find this link: https://wixtoolset.org/releases/v4-0-0-5205/
No idea if that's official, or some stuff a black hat guy shoved on an old unmaintained website (sorry I exaggerate).

Make a nice official release, or at least a clearly labeled preview or beta.  I don't want to spend time googling old github posts to find out if there is a thing like wix 4.0 or not.

**Reply by robmen** [reply]
Sure you can! I believe in you! [Here is a short video on how to get started with the very latest builds of WiX v4.0](https://youtu.be/2iIjq6zt6z0).

**Reply by glen-nicol** [reply]
Hi Rob, thanks for the link. I too have been pretty frustrated with WiX 4 documentation. It feels incoherent and disorganized. Even the video series you linked to has no numbering on the videos. Once I watched that one I had no idea which one is next. And I didn't see a link to any sort of index or playlist.

I understand WiX 4 is in progress and having to maintain documentation while changing things is annoying. But even if there was a single landing page with unorganized links to all the various videos and guides that have been written would be helpful so that maybe someone else can organize it.

Just some ideas. Thanks for the hard work!

EDIT: I was later able to find the [playlist for these videos](https://youtube.com/playlist?list=PLDlzbQXIs18slmqmdlS10_de_Cps-QRg6). YouTube didn't advertise it automatically in the watch next or related videos section.

**Reply by robmen** [reply]
> pretty frustrated with WiX 4 documentation.

What v4 documentation? Do you mean the [WIPs](https://wixtoolset.org/development/wips/)?

> maintain documentation while changing things is annoying.

Like the [WIPs](https://wixtoolset.org/development/wips/)? Or all the links to the weekly meetings in Discussion and the mailing lists that get summarized on https://www.firegiant.com/blog and linked to the Discussions?

> so that maybe someone else can organize it.

Yeah, but nobody does. Everyone just waits for the maintainers to do all of the work (there are a couple exceptions and those people are awesome).

Case in point, you never asked "I was trying to figure out X in WiX v4. I'd be happy to help improve doc about X if someone could help me understand X."

On top of that, I still don't even know what part of WiX v4 "documentation" you considered incoherent.

**Reply by glen-nicol** [reply]
Sorry for the delay.

This is the only [documentation](https://wixtoolset.org/documentation/manual/v4/reference/) I was aware of from the [main WiX toolset page](https://wixtoolset.org/documentation). I did all my WiX3 dev using the [WiX Toolset v3 Manual](https://wixtoolset.org/documentation/manual/v3/) and the great resource by [John Wright](https://wrightfully.com/series/writing-your-own-net-based-installer-with-wix). The v3 manual is fairly comprehensive and organized. The v4 page looks like a work in progress since it is listed as an equivalent to the v3 docs but isn't organized or as complete as the v3 page. I guess I expected it would have a sort of "start here for new projects" or "v4 migration guide" as a bare minimum explainer on how to get started.

I'll have to look through the WIP links. But I should mention that acronym is confusing. To me that is Work in Progress which seems like a promising lead to determine the state of v4 but then when I see it means Wix Improvement *proposals* that implies the contents are not completed or settled on. There is no indication of a resolution in the list.

I didn't want to come across as critical. In reality I am excited for v4. And I just wanted to mention that its been hard for me to determine if v4 is ready enough for production for us to even estimate time/effort required for a migration.  I think the root of my problem is lack of easily self-discoverable details to get started. Having to ask a question to get started is a high barrier of entry and seems like the answer for years in other threads has been "not ready yet". I never wanted to waste your time further.

**Reply by glen-nicol** [reply]
Was just checking back in for replies and realized I missed the ones below earlier (Sorry). I am happy to hear there is a single page guide in the works. And accept that the blog is the intended vector for v4 guides in the meantime. It may be moot now that the new guide is almost published but I would suggest linking to the blog from the wix toolset documentation page saying that is the best current resource.

**Reply by robmen** [reply]
There are some good things here, so I'm going to go through them in order (because it all builds):

>  The v4 page looks like a work in progress since it is listed as an equivalent

WiX v4 _is_ a work in progress. I don't think anyone is confused about that. Thus, any documentation about WiX v4 would _at best_ be a work in progress. Realistically speaking, there should be no expectations that user-facing documentation will be ready until `-rc.1`. If you want to use software before `-rc.1` it requires a deeper level of engagement in the project to be successful.

> I'll have to look through the WIP links. But I should mention that acronym is confusing.

Yeah... it would be best to read through at least the [first WIP](https://wixtoolset.org/development/wips/0000-wix-improvement-proposal/).

> There is no indication of a resolution in the list.

The workflow is described in the first WIP. I am considering moving WIPs to be wholly contained within the tracking issue instead of a separate document on the web. Either way, it doesn't really matter if they weren't read in the first place.

> I didn't want to come across as critical. 

Understood.

However, when providing feedback, it would be much appreciated if the feedback was actionable. For example, "I looked for status on WiX v4 in these locations: x, y, z. Maybe I'm blind, but I didn't see it. Where should I be looking to follow the status of WiX v4?" That provides useful data for us to consider, "Maybe we need to add links from locations x, y, and z to WiX v4 status."

Otherwise, it just feels critical. :)

>  if v4 is ready enough for production...

Every other week, there is a posting to this discussion list and the mailing list where the agenda of the meeting is sent out. One of the bullet items is regularly "status of v4". It sounds like that regular messaging missed you.

And to be clear, v4 isn't ready for production. When it is, we will push out announcements in all sorts of locations. We will try very, very hard to make it very, very obvious that v4 is ready for more use.

> Having to ask a question to get started is a high barrier of entry

Getting started with a project while it is under development requires significant activation energy. If you're not willing to ask questions to get started in the code, you're probably better off waiting for the announcement that the project is ready for more general use.

> seems like the answer for years in other threads has been "not ready yet"

Yes. And it was true. And we are working on it. And we're finally very close.

> blog from the wix toolset documentation page saying that is the best current resource.

Do you mean [this page](https://wixtoolset.org/documentation/)?

**Reply by glen-nicol** [reply]
Understood, maybe I got excited and ahead of myself when I saw a title that said "timeline" and then also see Rob say 

> You can use WiX v4.0 now.

Then saw the only docs that I was aware of hadn't changed and was confused/disappointed.

Yes that is the page I was thinking of as a root. But I would probably recommend linking from the [v4 page](https://wixtoolset.org/documentation/manual/v4/reference/) and change that to be more like the v3 page. 

I think what I had mind earlier about putting all your links on one page could be accomplished here simply by putting these things on a v4 specific page:
* Put a header that says in progress
* link to the current v4 schema pages that are there now
* a blurb about the WIPs and where to find them
* a blurb about the blog entries and where to find them

It doesn't have to be extensive. But having a landing page regardless of its completeness is really handy. But once again if you've just completed a guide that accomplished what's outlined above then don't worry about it and I'll wait a bit longer.

I'm excited to hear you're nearing the first version of wix4. : ) I bet that feels nice! And again sorry for contributing to any stressors.

**Comment by ChristianSauer** [comment]
I know this is hopeless but I will try anyway:

You try to sell / maintain a product. I you don't want to anymore that's ok, just announce it and be done.
But ATM I try to decide if I stay with WIX for a project I've taken over. WiX3 has some obvious aging problems, but Wix4 is just in a nirvana I don't have the time or energy to penetrate.
As a Developer I want a single page which explains how to use  WiX 4.0, maybe even how to upgrade from 3.0. Thats pretty standard in 2022 and your totally flabbergasted reaction suprises me.

**Reply by robmen** [reply]
> I know this is hopeless

It's only hopeless if you give up. The truth is out there.

> You try to sell / maintain a product.

I do [have a company that sells and maintains a couple of products](https://www.firegiant.com/products/). That company also [sells support services](https://www.firegiant.com/services/). But this here is an Open Source project. You don't get to treat it like a product.

>  but Wix4 is just in a nirvana I don't have the time or energy to penetrate.

You say you are interested in the Open Source WiX v4, but you don't want to take the time to follow a [blog that provides succinct highlights](https://www.firegiant.com/blog) of the status meeting every other week. That's a pretty low bar to follow for something you get for free.

> As a Developer I want a single page which explains how to use WiX 4.0, maybe even how to upgrade from 3.0. 

Sure. That's being (actually, I think it's just been) written.

> Thats pretty standard in 2022 and your totally flabbergasted reaction suprises me.

I'm "flabbergasted" (that is too strong a word really) that you want a page "which explains how to use WiX 4.0, maybe even how to upgrade from 3.0" when WiX v4 isn't (quite) ready for people to go through that process. We could write it but then they'd hit bugs we already know about and are fixing. If you want to use WiX v4 while it is under development (when bugs are being fixed), then you need to do extra work to understand the inner workings of the project here.

Honestly, it sounds like you want to treat this Open Source project like a product that you buy and get an SLA for support. There is a [company that does that for WiX today called FireGiant](https://www.firegiant.com/). You pay them for guarantees with a contract.

But the truth is this is an Open Source project. You get what you put in.

Hah, I knew we could find the truth!

**Comment by chrpai** [comment]
@ChristianSauer WiX v4 is very close to being ready for prime time.  I have an open source project that builds support v3 and v4.  I offer complimentary 1 hour consulting and I'd love to show you how easy it all can be.
